ocaml/xapi/repository_helpers.ml

```diff
@@ -169,7 +169,11 @@ let assert_url_is_valid ~url =
             ()
         | l -> (
           match
-            List.exists (fun d -> Astring.String.is_suffix ~affix:("." ^ d) h) l
+            List.exists
+              (fun d ->
+                Astring.String.is_suffix ~affix:("." ^ d) h || String.equal d h
+              )
+              l
           with
           | true ->
               ()
```

Comment:
I don't like all the nesting here I'd rather have a clause with a `when is_allowed domains`, in any case the way to do subdomain checking is to use [Domain_name.is_subdomain](https://hannesm.github.io/domain-name/doc/domain-name/Domain_name/index.html#val-is_subdomain)

```ocaml
let subdomain = Domain_name.host_exn h in 
let allow_hostname =
  List.exists (fun d -> Domain_name.(is_subdomain ~subdomain ~domain:(host_exn d)))
in
...
        | domains when not (allow_hostname domains) ->
            let msg = "host is not in allowlist" in
            raise Api_errors.(Server_error (internal_error, [msg]))
        | _ ->
            ()
```

Reply:
Unrelated, but `repository_domain_name_allowlist` is a poor name. This list is empty by default, which means nothing is allowed?

Reply:
I think that is indeed the case.

Reply:
`Domain_name.host_exn` is defined below:

Reply:
This is the kind of code we want to avoid. Addresses in IPv6 are looking different and this would not work. At the very least, code like this should be isolated in a module.

Reply:
The name of the allowd list states that those are domain names, IPs are not domain names. Whether this is the intention or not, I do not know, I'll refer to Ming to see what is the intent here

Reply:
In product, we should expect a `static` configuration is done always as something like "xenserver.com" in `repository_domain_name_allowlist`. This is done by another XS specific RPM package.
IPs are used only for private testing purpose. The empty `repository_domain_name_allowlist` means no such allowing test would be done at all.
So I think for the function of checking against allowlist, the IPs should not be supported. For using IPs in testing, the allowlist could be set as empty.

Reply:
@psafont The code tests says "TLD should not be all-numeric". TLD stands for Top Level Domain?  In any case, the code is testing for `exists is_letter (Array.get t 0)`; it this to detect IP addresses? In any case, an IPv6 address could contain letters. Altogether, this appears to me fragile.

Reply:
With the `Domain_name`, it is good to determine the sub domains. And the checking in `Domain_name` seems better to have as the allowlist is a built-in/static configuration (no API to set/get it). In other words, the built-in allowlist should not be written with ambiguities. And since it's not exposed as API, could XAPI just accept what `Domain_name` does currently?
And for the address in URL, the built-in allowlist can enforce the restriction on it.

Reply:
the RFC states that domain names are a bunch of labels separated by dots.
> Each node has a label, which is zero to 63 octets in length.  Brother
nodes may not have the same label
> When a user needs to type a domain name, the length of each label is
omitted and the labels are separated by dots (".").
> To simplify implementations, the total number of octets that represent a
domain name (i.e., the sum of all label octets and label lengths) is
limited to 255.

These properties should be all tested, otherwise we should open PRs with known-good code

Reply:
In order to support for checking ip host, the `Domain_name` module extended with `Ipaddr` lib.